diff --git a/dlls/ntdll/unix/virtual.c b/dlls/ntdll/unix/virtual.c
--- a/dlls/ntdll/unix/virtual.c
+++ b/dlls/ntdll/unix/virtual.c
@@ -442,8 +442,8 @@ static void mmap_add_reserved_area( void *addr, SIZE_T size )
     assert( !((UINT_PTR)addr & host_page_mask) );
     assert( !(size & host_page_mask) );

-    if (!((intptr_t)addr + size)) size--;  /* avoid wrap-around */
+    if (size >= host_page_size && !((intptr_t)addr + size)) size -= host_page_size;
     end = (char *)addr + size;

     LIST_FOR_EACH( ptr, &reserved_areas )
     {
@@ -494,8 +494,8 @@ static void mmap_remove_reserved_area( void *addr, SIZE_T size )
     assert( !((UINT_PTR)addr & host_page_mask) );
     assert( !(size & host_page_mask) );

-    if (!((intptr_t)addr + size)) size--;  /* avoid wrap-around */
+    if (size >= host_page_size && !((intptr_t)addr + size)) size -= host_page_size;

     ptr = list_head( &reserved_areas );
     /* find the first area covering address */
     while (ptr)
